src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs

```diff
@@ -86,6 +87,88 @@ public static ValueTask<T[]> WhenAll<T>(IEnumerable<ValueTask<T>> tasks)
+            CancellationToken cancellationToken)
+        {
+            if (func is null)
+                throw new ArgumentNullException(nameof(func));
+            if (transform is null)
+                throw new ArgumentNullException(nameof(transform));
+
+            var intermediateResult = func(arg, cancellationToken);
+            if (intermediateResult.IsCompletedSuccessfully)
+            {
+                // Synchronous fast path if 'func' completes synchronously
+                var result = intermediateResult.Result;
+                if (cancellationToken.IsCancellationRequested)
+                    return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+
+                return new ValueTask<TResult>(transform(result, arg));
+            }
+            else if (intermediateResult.IsCanceled && cancellationToken.IsCancellationRequested)
+            {
+                // Synchronous fast path if 'func' cancels synchronously
+                return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+            }
+            else
+            {
+                // Asynchronous fallback path
+                return UnwrapAndTransformAsync(intermediateResult, transform, arg, cancellationToken);
+            }
+
+            static ValueTask<TResult> UnwrapAndTransformAsync(ValueTask<TIntermediate> intermediateResult, Func<TIntermediate, TArg, TResult> transform, TArg arg, CancellationToken cancellationToken)
+            {
+                // Apply the transformation function once a result is available. The behavior depends on the final
+                // status of 'intermediateResult' and the 'cancellationToken'.
+                //
+                // | 'intermediateResult'       | 'cancellationToken' | Behavior                                 |
+                // | -------------------------- | ------------------- | ---------------------------------------- |
+                // | Ran to completion          | Not cancelled       | Apply transform                          |
+                // | Ran to completion          | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (matching token) | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (mismatch token) | Not cancelled       | Cancel result without applying transform |
+                // | Cancelled (mismatch token) | Cancelled           | Cancel result without applying transform |
+                // | Direct fault¹              | Not cancelled       | Fault result without applying transform  |
+                // | Direct fault¹              | Cancelled           | Cancel result without applying transform |
+                // | Indirect fault             | Not cancelled       | Fault result without applying transform  |
+                // | Indirect fault             | Cancelled           | Cancel result without applying transform |
```

Comment:
love this.

Reply:
what'st eh difference of Direct vs Indirect fault?

Reply:
Direct fault means `func` throws its exception directly (a non-async implementation that throws). Indirect fault means `func` completes to the point of returning a `ValueTask<TIntermediate>` instance which (immediately or eventually) transitions to the faulted state.

Reply:
gotcha.  can you add that comment?

Reply:
➡️ Expanded comment to cover this case